samza-test/src/main/java/org/apache/samza/test/framework/TestRunner.java

```diff
@@ -0,0 +1,343 @@
+   * {@link TestRunner} if not registered already. Then it creates the stream partitions with the registered System
+   * <p>
+   * @param stream represents the stream that is supposed to be configured with {@link TestRunner}
+   * @return calling instance of {@link TestRunner} with {@code stream} configured with it
+   */
+  public TestRunner addOutputStream(CollectionStream stream) {
+    Preconditions.checkNotNull(stream);
+    Preconditions.checkState(stream.getInitPartitions().size() >= 1);
+    registerSystem(stream.getSystemName());
+    CollectionStreamSystem system = systems.get(stream.getSystemName());
+    StreamSpec spec = new StreamSpec(stream.getStreamName(), stream.getStreamName(), system.getSystemName(), stream.getInitPartitions().size());
+    system
+        .getFactory()
+        .getAdmin(system.getSystemName(), new MapConfig(system.getSystemConfigs()))
+        .createStream(spec);
+    configs.putAll(stream.getStreamConfig());
+    return this;
+  }
+
+  /**
+   * Utility to run a test configured using TestRunner
+   */
+  public void run() {
+    Preconditions.checkState((app == null && taskClass != null) || (app != null && taskClass == null),
+        "TestRunner should run for Low Level Task api or High Level Application Api");
+    final LocalApplicationRunner runner = new LocalApplicationRunner(new MapConfig(configs));
+    if (app == null) {
+      runner.runTask();
+      runner.waitForFinish();
+    } else {
+      runner.run(app);
+      runner.waitForFinish();
+    }
+  }
+
+  /**
+   *
+   * @param stream represents {@link CollectionStream} whose current state of partitions is requested to be fetched
+   * @param timeout poll timeout in Ms
+   * @param <T> represents type of message
+   *
+   * @return a map key of which represents the {@code partitionId} and value represents the current state of the partition
+   *         i.e messages in the partition
+   * @throws InterruptedException Thrown when a blocking poll has been interrupted by another thread.
+   */
+  public static <T> Map<Integer, List<T>> consumeStream(CollectionStream stream, Integer timeout) throws InterruptedException {
+    Preconditions.checkNotNull(stream);
+    Preconditions.checkNotNull(stream.getSystemName());
+    Preconditions.checkNotNull(systems.containsKey(stream.getSystemName()));
+    String streamName = stream.getStreamName();
+    String systemName = stream.getSystemName();
+    Set<SystemStreamPartition> ssps = new HashSet<>();
+    Set<String> streamNames = new HashSet<>();
+    streamNames.add(streamName);
+    InMemorySystemFactory factory = systems.get(systemName).getFactory();
+    Map<String, SystemStreamMetadata> metadata =
+        factory.getAdmin(systemName, new MapConfig()).getSystemStreamMetadata(streamNames);
+    InMemorySystemConsumer consumer = (InMemorySystemConsumer) factory.getConsumer(systemName, null, null);
```

Comment:
You also need to be careful with tests that just run one after another (not necessarily in parallel). If one test does not properly clean up, then could it leave some extra data somewhere that a different test might read? That would be hard to debug, because you could get different test behaviors based on the order of the tests being run.